
# Using custom alignment files

While the `run-all-rpbp-instances` (and `run-rpbp-pipeline`) scripts are designed to handle all steps from translating the raw fastq files from the sequencer into a high-confidence list of translated ORFs, it is also possible to start the pipeline from any step. For example, the trimming, filtering and aligning steps could be handled using a different preprocessing strategy. The configuration file must be created as usual, but the `riboseq_samples` field only needs to contain the names of the samples. (The path can be left blank.)

Then, the files produced by the external processing pipelines must be placed at the appropriate location according to the names generated by Rp-Bp. In particular, depending on which steps of preprocessing have been performed, the files should be placed in the following locations:

* Trimmed and quality filtered reads
    * **trimmed and filtered reads**. `riboseq_data`/without-adapters/`sample-name`[.`note`].fastq.gz
    
* Reads not aligning to ribosomal sequences
    * **retained reads**. `riboseq_data`/without-rrna/`sample-name`[.`note`].fastq.gz
    
* Aligned reads
    * **sorted reads aligned to the genome**. `riboseq_data`/without-rrna-mapping/`sample-name`[.`note`].bam
    
    * **aligned reads which map uniquely to the genome**. A sorted bam file containing all alignments of reads to the genome with multimapping reads filtered out. `riboseq_data`/without-rrna-mapping/`sample-name`[.`note`]-unique.bam
    
The files can also be symlinks with the appropriate name. Please see the usage instructions for more details about the expected content of each file. Only the last file must be in the expected location. For example, if trimming, filtering and aligning has been performed, only the alignment files must be present. The pipeline will issue warning messages that the earlier files are missing, but it will begin as normal once it finds the, e.g., alignment files.


```python

```
